ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+underscore. A smart contract implements an interface if it supports
+all of the methods from that interface. A smart contract may implement
+zero, one, or multiple interfaces.
+
+#### Interface Description
+
+An interface description is a JSON object containing the JSON
+descriptions for each of the methods in the interface.
+
+The JSON structure for such an object is:
+
+```
+interface Interface {
+  name: string,
+  methods: Array<Method>
+}
+```
+
+For example:
+
+```
+{
+  "name": "Calculator",
+  "methods": [
+    { "name": "add", "args": [...], ... },
+    { "name": "multiply", "args": [...], ... }
+  ]
+}
+```
+
+### Contracts
+
+An contract is the complete set of methods that an app implements. It
+is similar to an interface, but may include further details about the
+concrete implementation. All methods in an contract must be unique
+(specifically, they must have unique method selectors). Method names
+in Contracts **MAY** begin with underscore, but these names are
+reserved for use by this, or suceeding specifications.
+
+A contract may contain two special methods, `_optIn` and
+`_closeOut`. These methods describe how contracts should opt-in, or
+close-out such apps.  They may have any arguments or return values,
+but there must be no more than one method with each name.
+
+If either of these methods is elided, an ARC-4 app **MUST** allow
+opt-in or close-out transactions with no arguments.
+
+#### Contract Description
+
+An contract description is a JSON object containing the JSON
+descriptions for each of the methods in the contract.
+
+The JSON structure for such an object is:
+
+```
+interface Contract {
+  name: string,
+  appId: number
+  methods: Array<Method>
+}
+```
+
+For example:
+
+```
+{
+  "name": "Calculator",
+  "appId": 1762763,
+  "methods": [
+    { "name": "add", "args": [...], ... },
+    { "name": "multiply", "args": [...], ... }
+  ]
+}
+```
+
+
+## Method Invocation
+
+In order for a caller to invoke a method, the caller and the method
+implementation (callee) must agree on how information will be passed
+to and from the method. This ABI defines a standard for where this
+information should be stored and its format.
+
+
+### Standard Format
+
+The method selector will be the first application call argument,
+accessible as `txn ApplicationArgs 0` from TEAL. For the two special
+methods `_optIn` and `_closeOut`, the first argument **MUST** be
+empty. Instead, the OnCompletion (`apan`) field of the application
+call transaction is set appropriately. For all other methods, `apan`
+is set to NoOpOC (`0`).
+
+The first 14 method arguments will occupy the next 14 application call
+arguments, accessible from TEAL as `txn ApplicationArgs i` for `1 <= i
+<= 14`. The arguments will be encoded as defined in the Encoding
+section.
+
+If a method has more than 14 arguments, the remaining arguments will
+be encoded as a tuple in the final application call argument,
+accessible from TEAL as `txn ApplicationArgs 15`.
```

Comment:
`txn ApplicationArgs 15` is assembled the same as `txna ApplicationArgs 15` because we pick the opcode based on the number of immediates.  Is that the typo you mean?

Tell me what `txnaa` would do.  I was thinking maybe you just want `txnas` which would get the array index from stack.  Then you could do your 14 loads in a short loop.

The path you described would need all the typing info, wouldn't it? But it's not present during execution.

Reply:
This shorthand (`txn` is sometimes `txna`) is not documented anywhere that I can find and I have never seen it before. Given that, I think it is a bad idea to use it in documentation when `txna` is more explicit. That said... you should add it to the documentation.

`txnaa` is `txna ApplicationArgs` but is 2 bytes rather than 3, so there's no incentive to put them in scratch slots.

Yea, `getABI` needs type info, so just ignore that.

Reply:
I agree we should probably change instances of `txn ApplicationArgs X` to use `txna` because that syntax is more common

Reply:
Agreed AND document the other syntax --- https://developer.algorand.org/docs/reference/teal/opcodes/#txn-f